### PR TITLE
Key attribute support on Primary Key field when using POCO classes with the client

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes issue #1332.*

### Description
OData client supports 3 ways of defining an entity key:
1. ID
2. type+ID
3. KeyAttribute
 Currently, the client does not support the use of the Key  attribute annotation on the client when using POCO classes.  Like below: 
<pre>
//POCO class
public class ProductDetails
{
    [Key]
    public string NonStandardId { get; set; }
    public string OtherField { get; set; }
}</pre>
This PR adds support for the use of the Key attribute on a primary key field. 
### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 1]
```diff
@@ -426,26 +429,44 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
...
+                if (context != null)
+                {
+                    IEdmModel model = context.Format.ServiceModel;
+                    if (model != null && model is ClientEdmModel)
+                    {
+                        EdmEntityType edmEntityType = (EdmEntityType)model.SchemaElements.FirstOrDefault();
+                        if (edmEntityType != null)
+                        {
+                            var declaredKey = edmEntityType.DeclaredKey;
+                            if (declaredKey != null)
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: These tests can be simplified with to reduce the nesting maybe having
```if (context?.Format.ServiceModel is ClientEdmModel model)``` maybe followed by ``` model.SchemaElements.FirstOrDefault()?.DeclaredKey...```

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 2]
```diff
@@ -426,26 +429,44 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
...
             List<PropertyInfo> keyProperties = new List<PropertyInfo>();
             PropertyInfo[] properties = ClientTypeUtil.GetPropertiesOnType(type, false /*declaredOnly*/).ToArray();
             hasProperties = properties.Length > 0;
-
             KeyKind currentKeyKind = KeyKind.NotKey;
             KeyKind newKeyKind = KeyKind.NotKey;
             foreach (PropertyInfo propertyInfo in properties)
             {
-                if ((newKeyKind = ClientTypeUtil.IsKeyProperty(propertyInfo, dataServiceKeyAttribute)) != KeyKind.NotKey)
+                if ((newKeyKind = ClientTypeUtil.IsKeyProperty(propertyInfo, dataServiceKeyAttribute,isKeyDeclared)) != KeyKind.NotKey)
```
Comment (g2mula) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]:
```suggestion
                if ((newKeyKind = ClientTypeUtil.IsKeyProperty(propertyInfo, dataServiceKeyAttribute, isKeyDeclared)) != KeyKind.NotKey)
```
Whitespace

File: test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
+﻿//---------------------------------------------------------------------
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DataServiceContextMethodMockSampleTest.cs]: Maybe this checkin is meant for a different PR? It doesn't seem related to #1332

File: src/Microsoft.OData.Client/ALinq/ExpressionNormalizer.cs [hunk 4]
```diff
@@ -174,7 +188,7 @@ internal override Expression VisitUnary(UnaryExpression u)
                 if (!PrimitiveType.IsKnownNullableType(visited.Operand.Type) && !PrimitiveType.IsKnownNullableType(visited.Type) || visited.Operand.Type == visited.Type)
                 {
                     // x is not a collection of entity types
-                    if(!(ClientTypeUtil.TypeOrElementTypeIsEntity(visited.Operand.Type) && ProjectionAnalyzer.IsCollectionProducingExpression(visited.Operand)))
+                    if(!(ClientTypeUtil.TypeOrElementTypeIsEntity(visited.Operand.Type,this.context) && ProjectionAnalyzer.IsCollectionProducingExpression(visited.Operand)))
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/ALinq/ExpressionNormalizer.cs]: Fixed

File: src/Microsoft.OData.Client/ALinq/ResourceExpression.cs [hunk 5]
```diff
@@ -94,6 +94,10 @@ internal ResourceExpression(Expression source, Type type, List<string> expandPat
         {
         }
 
+        internal ResourceExpression(DataServiceContext context)
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/ALinq/ResourceExpression.cs]: Fixed

File: src/Microsoft.OData.Client/ALinq/ResourceExpression.cs [hunk 6]
```diff
@@ -123,6 +127,11 @@ internal ResourceExpression(Expression source, Type type, List<string> expandPat
             this.isAction = isAction;
         }
 
+        //internal ResourceExpression(DataServiceContext context)
+        //{
+        //    this.context = context;
+        //}
+
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/ALinq/ResourceExpression.cs]: Fixed

File: src/Microsoft.OData.Client/DataServiceQueryOfT.cs [hunk 7]
```diff
@@ -114,7 +114,7 @@ internal set
         }
 
         /// <summary>Context associated with this query.</summary>
-        public DataServiceContext Context
+        public virtual DataServiceContext Context
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/DataServiceQueryOfT.cs]: Fixed

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 8]
```diff
@@ -426,26 +429,44 @@ internal static PropertyInfo[] GetKeyPropertiesOnType(Type type)
...
         {
             if (CommonUtil.IsUnsupportedType(type))
             {
                 throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
             }
-
             string typeName = type.ToString();
             IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+            bool isKeyDeclared = false;
+            if (customAttributes.Count() == 0)
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: Fixed

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs (hunk 1 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: Fixed

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs (hunk 2 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: Fixed

File: test/FunctionalTests/Microsoft.OData.Client.Tests/packages.config [hunk 9]
```diff
@@ -1,5 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Castle.Core" version="4.4.0" targetFramework="net45" />
+  <package id="Moq" version="4.13.1" targetFramework="net45" />
+  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.0" targetFramework="net45" />
+  <package id="System.Threading.Tasks.Extensions" version="4.5.1" targetFramework="net45" />
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/packages.config]: Maybe this was not meant for checkin for this particular PR?

File: test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/Materialization/EntryValueMaterializationPolicyUnitTests.cs [hunk 10]
```diff
@@ -301,7 +301,8 @@ private IEnumerable<TestOrder> GetOverlappingOrders(IEnumerable<TestOrder> exist
                 materializerContext,
                 adapter,
                 null,
-                new Dictionary<IEnumerable, DataServiceQueryContinuation>());
+                new Dictionary<IEnumerable, DataServiceQueryContinuation>()
+                );
```
Comment (g2mula) [inline test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/Materialization/EntryValueMaterializationPolicyUnitTests.cs]: Probably wasn't meant for checkin?

File: test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.csproj [hunk 11]
```diff
@@ -41,11 +42,24 @@
...
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="Castle.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=407dd0808d44fbdc, processorArchitecture=MSIL">
+      <HintPath>..\..\..\sln\packages\Castle.Core.4.4.0\lib\net45\Castle.Core.dll</HintPath>
+    </Reference>
     <Reference Include="FluentAssertions, Version=4.1.0.0, Culture=neutral, PublicKeyToken=33f2691a05b67b6a, processorArchitecture=MSIL">
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\..\sln\packages\FluentAssertions.2.0.0.1\lib\net40\FluentAssertions.dll</HintPath>
     </Reference>
+    <Reference Include="Moq, Version=4.13.0.0, Culture=neutral, PublicKeyToken=69f491c39445e920, processorArchitecture=MSIL">
```
Comment (g2mula) [inline test/FunctionalTests/Microsoft.OData.Client.Tests/Microsoft.OData.Client.Tests.csproj]: Probably wasn't meant for checkin for this particular PR?

File: src/Microsoft.OData.Client/ClientEdmModel.cs [hunk 12]
```diff
@@ -61,6 +71,10 @@ internal ClientEdmModel(ODataProtocolVersion maxProtocolVersion)
             this.maxProtocolVersion = maxProtocolVersion;
             this.EdmStructuredSchemaElements = new ConcurrentEdmSchemaDictionary();
         }
+        internal ClientEdmModel(DataServiceContext context)
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.OData.Client/ClientEdmModel.cs]: +1 and comments

---
In reply to: [380839161](https://github.com/OData/odata.net/pull/1663#discussion_r380839161) [](ancestors = 380839161)

Comment (xuzhg) [inline src/Microsoft.OData.Client/ClientEdmModel.cs]: >ClientEdmModel(DataServiceContext context) [](start = 17, length = 42)

inject a "DataServiceContext" into "ClientEdmModel" is not good design. the Model should not internally depend on the context itself. However, we can get the EdmModel using the context. Right?

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 13]
```diff
@@ -413,39 +417,58 @@ internal static IEnumerable<PropertyInfo> GetPropertiesOnType(Type type, bool de
...
 
         /// <summary>
         /// Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.
         /// </summary>
         /// <param name="type">Type in question.</param>
+        /// <param name="context">The context with the client model</param>
         /// <param name="hasProperties">true if <paramref name="type"/> has any (declared or inherited) properties; otherwise false.</param>
         /// <returns>Returns the list of key properties defined on <paramref name="type"/>; null if <paramref name="type"/> is complex.</returns>
-        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, out bool hasProperties)
+        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, DataServiceContext context, out bool hasProperties)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: >DataServiceContext context, [](start = 73, length = 27)

don't inject a Context into this static method.

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 14]
```diff
@@ -693,7 +713,7 @@ internal static string GetEnumValuesString(string enumString, Type enumType)
         /// <param name="propertyInfo">Property in question.</param>
         /// <param name="dataServiceKeyAttribute">DataServiceKeyAttribute instance.</param>
         /// <returns>Returns the KeyKind if <paramref name="propertyInfo"/> is declared as a key in <paramref name="dataServiceKeyAttribute"/> or it follows the key naming convention.</returns>
-        private static KeyKind IsKeyProperty(PropertyInfo propertyInfo, KeyAttribute dataServiceKeyAttribute)
+        private static KeyKind IsKeyProperty(PropertyInfo propertyInfo, KeyAttribute dataServiceKeyAttribute, List<string> declaredKeys)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: >List<string> [](start = 110, length = 12)

why not use "IList<PropertyIno> declaredKeys"?

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 15]
```diff
@@ -718,10 +738,12 @@ private static KeyKind IsKeyProperty(PropertyInfo propertyInfo, KeyAttribute dat
                     keyKind = KeyKind.Id;
                 }
             }
-
+            else if (declaredKeys != null && declaredKeys.Contains(propertyName))
+            {
+                keyKind = KeyKind.DeclaredKey;
+            }
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: I think we should move these codes before the  "convention", right?

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 16]
```diff
@@ -413,39 +417,58 @@ internal static IEnumerable<PropertyInfo> GetPropertiesOnType(Type type, bool de
...
         {
             if (CommonUtil.IsUnsupportedType(type))
             {
                 throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
             }
-
             string typeName = type.ToString();
             IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+            List<string> DeclaredKeys = new List<string>();
+            if (!customAttributes.Any())
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: >f (!customAttributes.Any()) [](start = 13, length = 27)

why test attribute if you have the model to retrieve the keys?

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 17]
```diff
@@ -413,39 +417,58 @@ internal static IEnumerable<PropertyInfo> GetPropertiesOnType(Type type, bool de
...
+            {
+                if (context?.Format.ServiceModel is EdmModel)
+                {
+                    var model = context.Format.ServiceModel as EdmModel;
+                    if (model?.SchemaElements.FirstOrDefault() is EdmEntityType)
+                    {
+                        var edmEntityType = model.SchemaElements.FirstOrDefault() as EdmEntityType;
+                        if (edmEntityType?.DeclaredKey != null)
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: >DeclaredKey [](start = 43, length = 11)

Should use "key()" to include the keys defined on the base type?

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs [hunk 18]
```diff
@@ -413,39 +417,58 @@ internal static IEnumerable<PropertyInfo> GetPropertiesOnType(Type type, bool de
...
+        internal static PropertyInfo[] GetKeyPropertiesOnType(Type type, DataServiceContext context, out bool hasProperties)
         {
             if (CommonUtil.IsUnsupportedType(type))
             {
                 throw new InvalidOperationException(c.Strings.ClientType_UnsupportedType(type));
             }
-
             string typeName = type.ToString();
             IEnumerable<object> customAttributes = type.GetCustomAttributes(true);
+            List<string> DeclaredKeys = new List<string>();
```
Comment (xuzhg) [inline src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: >DeclaredKeys [](start = 25, length = 12)

don't upper initial for the local variable

File: src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs (hunk 16 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs]: The first check checks whether one used the Key Attribute annotation on the key property when creating their model, If not then get the keys from the model.
